app/javascript/controllers/pathogen/tabs_controller.js

```diff
@@ -0,0 +1,641 @@
+      this.element.innerHTML =
+        '<div class="text-red-600">At least one panel target is required</div>';
+      return false;
+    }
+
+    if (this.tabTargets.length !== this.panelTargets.length) {
+      console.error("[pathogen--tabs] Tab and panel counts must match", {
+        tabs: this.tabTargets.length,
+        panels: this.panelTargets.length,
+      });
+      this.element.innerHTML =
+        '<div class="text-red-600">Tab and panel counts must match</div>';
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Validates and normalizes the default index value
+   * @private
+   * @param {number} index - The default index
+   * @returns {number} Validated index (0 if invalid)
+   */
+  #validateDefaultIndex(index) {
+    if (index < 0 || index >= this.tabTargets.length) {
+      console.warn(
+        `[pathogen--tabs] default_index ${index} out of bounds, using 0`,
+      );
+      return 0;
+    }
+    return index;
+  }
+
+  /**
+   * Sets up ARIA attributes on tabs and panels
+   * @private
+   * @returns {void}
+   */
+  #setupARIA() {
+    this.tabTargets.forEach((tab, index) => {
+      const panel = this.panelTargets[index];
+
+      // Ensure tab has required ARIA attributes
+      if (!tab.hasAttribute("role")) {
+        tab.setAttribute("role", "tab");
+      }
+
+      // Link tab to panel
+      if (panel && panel.id) {
+        tab.setAttribute("aria-controls", panel.id);
+      }
+
+      // Initial aria-selected (will be updated by selectTabByIndex)
+      tab.setAttribute("aria-selected", "false");
+
+      // Initial tabindex (will be updated by selectTabByIndex)
+      tab.tabIndex = -1;
+    });
+
+    this.panelTargets.forEach((panel, index) => {
+      const tab = this.tabTargets[index];
+
+      // Ensure panel has required ARIA attributes
+      if (!panel.hasAttribute("role")) {
+        panel.setAttribute("role", "tabpanel");
+      }
+
+      // Link panel to tab
+      if (tab && tab.id) {
+        panel.setAttribute("aria-labelledby", tab.id);
+      }
+
+      // Initial hidden state (will be updated by selectTabByIndex)
+      panel.setAttribute("aria-hidden", "true");
+    });
+  }
+
+  /**
+   * Selects a tab by index
+   *
+   * This method handles both tab selection state and panel visibility.
+   * When a panel becomes visible (hidden class removed), any Turbo Frames
+   * with loading="lazy" inside will automatically fetch their content.
+   *
+   * Turbo Frame Integration:
+   * - Removing the 'hidden' class triggers Turbo's lazy loading mechanism
+   * - Turbo automatically fetches frame content when it becomes visible
+   * - Once loaded, Turbo caches the content (no refetch on revisit)
+   * - No explicit fetch() or JavaScript handling needed
+   *
+   * @private
+   * @param {number} index - The tab index to select
+   * @param {boolean} updateUrl - Whether to update the URL hash (default: true)
+   * @returns {void}
+   */
+  #selectTabByIndex(index, updateUrl = true) {
```

Comment:
Suggestion (blocking): Add an additional argument here to set the history update method, so that when you call `this.#updateUrlHash` you can pass the update method e.g. `history.replaceState` when this method is called by connect and `history.pushState` otherwise.